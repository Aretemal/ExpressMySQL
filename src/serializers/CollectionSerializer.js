import fullUrlCreator from '../utils/fullUrlCreator.js';
import Serializer from './Serializer.js';

class CollectionSerializer extends Serializer {
  constructor(resource, serializerType, request = null, metaData = null) {
    super(resource, request);
    this.serializerType = serializerType;
    this.metaData = metaData;
  }

  meta() {
    return this.metaData;
  }

  serialize() {
    const data = {
      data: this.collect(),
      links: this.links(),
    };
    if (this.metaData) {
      data.meta = this.meta();
    }
    return data;
  }

  collect() {
    return this.resource.map((item) => {
      const serializer = new this.serializerType(item);
      return serializer.serialize().data;
    });
  }

  links() {
    return { self: `${process.env.API_URL}/${this.request.originalUrl}` };
  }
}
export default CollectionSerializer;
