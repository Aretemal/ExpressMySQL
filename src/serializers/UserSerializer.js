import fullUrlCreator from '../utils/fullUrlCreator.js';
import Serializer from './Serializer.js';

class UserSerializer extends Serializer {
  type() {
    return 'User';
  }

  attributes() {
    return {
      ava: this.resource.ava,
      email: this.resource.email,
      firstName: this.resource.firstName,
      lastName: this.resource.lastName,
      login: this.resource.login,
      status: this.resource.status,
    };
  }

  attributes() {
    return {
      ...this.resource,
    };
  }

  links() {
    return { self: `${process.env.API_URL}/user/${this.resource.id}` };
  }
}
export default UserSerializer;
