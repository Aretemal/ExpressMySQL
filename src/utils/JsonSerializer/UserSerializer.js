import Serializer from './Serializer.js';

class UserSerializer extends Serializer {
  getOne(data, links) {
    return this.serializer(data, data.id, 'ObjectUser', links);
  }

  getAllUsers(data, links) {
    return this.serializer(data, data.userAuth.id, 'Array Users', links);
  }

  getStatus(data, links) {
    return this.serializer(data, data.id, 'Status', links);
  }
}
export default new UserSerializer();
