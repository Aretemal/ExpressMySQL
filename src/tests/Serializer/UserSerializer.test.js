import { matchersWithOptions } from 'jest-json-schema';
import CollectionSerializer from '../../serializers/CollectionSerializer.js';
import UserSerializer from '../../serializers/UserSerializer.js';
import schema from './schema.json';

expect.extend(matchersWithOptions({
  verbose: true,
}));

describe('UserSerializer :', () => {
  describe('getOne :', () => {
    test('should return info about user', () => {
      const user = {
        id: 1,
        login: 'Artem',
        firstName: 'aaaa',
        lastName: 'bbbb',
        email: '111@11.1',
        status: '1',
        ava: null,
      };

      const serializer = new UserSerializer(user);
      const data = serializer.serialize();

      expect(data).toMatchSnapshot();
      expect(data).toMatchSchema(schema);
    });
  });

  describe('getAllUsers :', () => {
    test('should return array with info', () => {
      const countOfUsers = 2;
      const users = [{
        id: 3,
        login: 'dsdsd',
        firstName: 'assaaa',
        lastName: 'bbsbb',
        email: '111s@11.1',
        status: '1s',
        ava: null,
      },
      {
        id: 1,
        login: 'Artem',
        firstName: 'aaaa',
        lastName: 'bbbb',
        email: '111@11.1',
        status: '1',
        ava: null,
      }];

      const serializer = new CollectionSerializer(
        users,
        {
          serializerType: UserSerializer,
          metaData: { countOfUsers },
        },
      );
      const data = serializer.serialize({ originalUrl: 'example' });

      expect(data).toMatchSnapshot();
      expect(data).toMatchSchema(schema);
    });
  });
});
