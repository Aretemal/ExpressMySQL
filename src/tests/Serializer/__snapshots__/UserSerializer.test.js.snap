// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSerializer : getAllUsers : should return array with info 1`] = `
{
  "data": [
    {
      "attributes": {
        "ava": null,
        "email": "111s@11.1",
        "firstName": "assaaa",
        "lastName": "bbsbb",
        "login": "dsdsd",
        "status": "1s",
      },
      "id": "3",
      "links": {
        "self": "http://localhost:5000/api/user/3",
      },
      "type": "User",
    },
    {
      "attributes": {
        "ava": null,
        "email": "111@11.1",
        "firstName": "aaaa",
        "lastName": "bbbb",
        "login": "Artem",
        "status": "1",
      },
      "id": "1",
      "links": {
        "self": "http://localhost:5000/api/user/1",
      },
      "type": "User",
    },
  ],
  "links": {
    "self": "http://localhost:5000/api/example",
  },
  "meta": {
    "countOfUsers": 2,
  },
}
`;

exports[`UserSerializer : getOne : should return info about user 1`] = `
{
  "data": {
    "attributes": {
      "ava": null,
      "email": "111@11.1",
      "firstName": "aaaa",
      "lastName": "bbbb",
      "login": "Artem",
      "status": "1",
    },
    "id": "1",
    "links": {
      "self": "http://localhost:5000/api/user/1",
    },
    "type": "User",
  },
}
`;
